APE26.rst

```diff
@@ -0,0 +1,220 @@
+and the storage of coordinate data within that frame. Ideally the code would be more 
+modularly structured, where the coordinate frame class **only** defines the reference 
+frame. It would then be the purview of another class – `SkyCoord` – to bring these 
+concerns together and to represent data (using `Representation` and `Differential` 
+classes) in a given reference frame (using a `CoordinateFrame` class). As a demonstration 
+of the current state of duplicated functionality, these two initializations effectively 
+represent the same thing, but return different objects with similar but not identical 
+APIs:
+
+.. code-block:: python
+
+    c1 = SkyCoord(1., 2., frame="icrs", units="deg")
+    c2 = ICRS(1. * u.deg, 2. * u.deg)
+
+We suggest the situation should instead be analogous to what is the case for units, 
+which know how to transform from one to another, but are not concerned with how the 
+values are stored (that belongs to ``Quantity``). Translating to coordinates, units are 
+like ``CoordinateFrame``, the values are the coordinate data (``Representations``), and 
+``Quantity`` is like ``SkyCoord``.
+
+Having both the frame classes as well as ``SkyCoord`` be able to store and handle data 
+has resulted in a large amount of code duplication. Restructuring the ``Frame`` classes 
+to remove data storage will allow for much more maintainable, de-duplicated code. It 
+will also make it easier to contribute: if there is a problem one would like to solve 
+in a given method, if one looks in ``SkyCoord``, one will likely find that it does not 
+exist, but instead is defined on ``BaseCoordinateFrame`` and gets dynamically called via 
+``SkyCoord.__getattr__``. Indeed, the construction of ``BaseCoordinateFrame`` ends up 
+complicating ``SkyCoord``, which has to manage the coordinate data **through** the stored 
+``BaseCoordinateFrame``.
+
+Another issue with the current implementation of coordinate frames is that the optional 
+inclusion of coordinate data makes the reference frames “multi-modal”. This creates 
+different usage modes (with and without data), each exhibiting different behavior. For 
+instance, some methods such as ``separation`` work fine with frames with data, while 
+doing this on coordinate frames without data results in an error. This kind of 
+multi-modal structure is considered an anti-pattern because it forces any code 
+interacting with the frame classes to handle both cases (checking ``frame.has_data``), 
+complicating the codebase unnecessarily.
+
+All the points discussed thus far – separation of concerns and code duplication – 
+concern maintainers. However user experience is the more important consideration. In 
+this arena too, separating frames from data storage has its advantages. Perhaps most 
+importantly, documentation will be more obvious: the methods and attributes are defined 
+on ``SkyCoord``, and sphinx will know how to typeset those. It will also be easier: 
```

Comment:
This is relevant not only for documentation but also for type annotations. Looking up `BaseCoordinateFrame` methods via `SkyCoord.__getattr__()` is too dynamic for static type checkers. It is possible to partially alleviate this by writing out stubs in `SkyCoord` (e.g. https://github.com/astropy/astropy/blob/24d48a9e2ef90af5383bf76e6d098e6d193d8186/astropy/coordinates/sky_coordinate.py#L178-L181), but type checkers cannot ensure that the stubs agree with the signatures of the actual methods.